### PR TITLE
Setup Github Actions

### LINKED ISSUE #1620: Travis-CI.org is shutdown
I could not find zim-desktop-wiki at Travis CI:
- https://app.travis-ci.com/search/zim-desktop-wiki returns nothing
- https://app.travis-ci.com/github/zim-desktop-wiki/zim-desktop-wiki/ does not work

Is it gone with https://travis-ci.org/?

> Since June 15th, 2021, the building on travis-ci.org is ceased. Please use travis-ci.com from now on.

https://blog.travis-ci.com/2021-05-07-orgshutdown

### PR DESCRIPTION
**Pull request by pdecat**
This PR sets up Github Actions as Travis-CI.org is shutdown.

Resolves #1620.

### CODE REVIEW

**Comment by pdecat** [comment]
I've opened a local PR on my own fork to actually get test runs: https://github.com/pdecat/zim-desktop-wiki/pull/1/checks

**Comment by pdecat** [comment]
@jaap-karssenberg CI is green in my fork :tada: https://github.com/pdecat/zim-desktop-wiki/pull/1

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      - name: Install python dependencies
+        run: |
+          pip install PyGObject
+          pip install pyxdg
+      - name: Configure git
+        run: |
+          git config --global user.email "you@example.com"
+          git config --global user.name "Your Name"
+      - name: Test
+        run: LC_COLLATE="en_US.UTF-8" xvfb-run ./test.py
```
Comment (pdecat) [inline .github/workflows/test.yml:39]: `LC_COLLATE="en_US.UTF-8"` is required to fix the following error:

```
======================================================================
FAIL: runTest (tests.widgets.TestLinkEntry)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/widgets.py", line 254, in runTest
    TestPageEntry.runTest(self)
  File "/home/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/widgets.py", line 229, in runTest
    self.assertEqual([r[0] for r in self.completion_model], wanted)
AssertionError: Lists differ: [':žžž', ':Bar', ':Maßnahmen', ':Placeholder', ':Test'] != [':Bar', ':Maßnahmen', ':Placeholder', ':Test', ':žžž']

First differing element 0:
':žžž'
':Bar'

- [':žžž', ':Bar', ':Maßnahmen', ':Placeholder', ':Test']
?  --------

+ [':Bar', ':Maßnahmen', ':Placeholder', ':Test', ':žžž']
?                                               ++++++++


======================================================================
FAIL: runTest (tests.widgets.TestNamespaceEntry)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/widgets.py", line 245, in runTest
    TestPageEntry.runTest(self)
  File "/home/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/widgets.py", line 229, in runTest
    self.assertEqual([r[0] for r in self.completion_model], wanted)
AssertionError: Lists differ: [':žžž', ':Bar', ':Maßnahmen', ':Placeholder', ':Test'] != [':Bar', ':Maßnahmen', ':Placeholder', ':Test', ':žžž']

First differing element 0:
':žžž'
':Bar'

- [':žžž', ':Bar', ':Maßnahmen', ':Placeholder', ':Test']
?  --------

+ [':Bar', ':Maßnahmen', ':Placeholder', ':Test', ':žžž']
?                                               ++++++++


======================================================================
FAIL: runTest (tests.widgets.TestPageEntry)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/widgets.py", line 229, in runTest
    self.assertEqual([r[0] for r in self.completion_model], wanted)
AssertionError: Lists differ: [':žžž', ':Bar', ':Maßnahmen', ':Placeholder', ':Test'] != [':Bar', ':Maßnahmen', ':Placeholder', ':Test', ':žžž']

First differing element 0:
':žžž'
':Bar'

- [':žžž', ':Bar', ':Maßnahmen', ':Placeholder', ':Test']
?  --------

+ [':Bar', ':Maßnahmen', ':Placeholder', ':Test', ':žžž']
?
```

See https://github.com/pdecat/zim-desktop-wiki/pull/1/checks?check_run_id=3708067156#step:7:995

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+          sudo apt-get update -qq
+          sudo apt-get install -qq -y gir1.2-gtk-3.0 gobject-introspection libgirepository1.0-dev xvfb
+      - name: Install python dependencies
+        run: |
+          pip install PyGObject
+          pip install pyxdg
+      - name: Configure git
+        run: |
+          git config --global user.email "you@example.com"
+          git config --global user.name "Your Name"
```
Comment (pdecat) [inline .github/workflows/test.yml:37]: This is required to fix the following error:

```
ERROR: runTest (tests.versioncontrol.TestGit)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/versioncontrol.py", line 339, in runTest
    vcs.commit_version('test 1')
  File "/home/runner/work/zim-desktop-wiki/zim-desktop-wiki/zim/plugins/versioncontrol/__init__.py", line 521, in commit_version
    self.commit(None, msg)
  File "/home/runner/work/zim-desktop-wiki/zim-desktop-wiki/zim/plugins/versioncontrol/git.py", line 109, in commit
    return self.run(params)
  File "/home/runner/work/zim-desktop-wiki/zim-desktop-wiki/zim/plugins/versioncontrol/__init__.py", line 451, in run
    self._app.run(params, self.root)
  File "/home/runner/work/zim-desktop-wiki/zim-desktop-wiki/zim/applications.py", line 269, in run
    raise ApplicationError(argv[0], argv[1:], p.returncode, _decode(stderr))
zim.applications.ApplicationError: Failed to run application: git

git "commit" "-m" "test 1"
returned non-zero exit status 128

Author identity unknown

*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: empty ident name (for <runner@fv-az190-546.22yfuiovmbsufjzmhpaxkig25b.bx.internal.cloudapp.net>) not allowed
```

See https://github.com/pdecat/zim-desktop-wiki/runs/3707947920?check_suite_focus=true#step:6:1030

**Comment by pdecat** [comment]
There's a `Could not run desktop-file-validate` message at the end, not sure if it matters: https://github.com/pdecat/zim-desktop-wiki/runs/3708160093?check_suite_focus=true#step:7:998

**Comment by pdecat** [comment]
Tests are now also run Windows and MacOS, but some are failing.

# Windows

```
======================================================================
FAIL: runTest (tests.notebook.TestNotebookInfoList)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:/a\zim-desktop-wiki\zim-desktop-wiki/tests/notebook.py", line 138, in runTest
    self.assertEqual(list[:], [
AssertionError: Lists differ: [<Not[85 chars]e:///D:/home/user/code/zim.debug>, <NotebookIn[31 chars]Bar>] != [<Not[85 chars]e:///C:/home/user/code/zim.debug>, <NotebookIn[31 chars]Bar>]

First differing element 1:
<NotebookInfo: file:///D:/home/user/code/zim.debug>
<NotebookInfo: file:///C:/home/user/code/zim.debug>

  [<NotebookInfo: file:///D:/a/_temp/msys/msys64/home/runneradmin/Notes>,
-  <NotebookInfo: file:///D:/home/user/code/zim.debug>,
?                         ^

+  <NotebookInfo: file:///C:/home/user/code/zim.debug>,
?                         ^

-  <NotebookInfo: file:///D:/home/user/Foo%20Bar>]
?                         ^

+  <NotebookInfo: file:///C:/home/user/Foo%20Bar>]
?                         ^


======================================================================
FAIL: runTest (tests.notebook.TestResolveNotebook)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:/a\zim-desktop-wiki\zim-desktop-wiki/tests/notebook.py", line 167, in runTest
    self.assertEqual(info.uri, uri)
AssertionError: 'file:///D:/a/_temp/msys/msys64/home/runneradmin/bar' != 'file:///C:/D:/a/_temp/msys/msys64/home/runneradmin/bar'
- file:///D:/a/_temp/msys/msys64/home/runneradmin/bar
+ file:///C:/D:/a/_temp/msys/msys64/home/runneradmin/bar
?        +++


======================================================================
FAIL: runTest (tests.widgets.TestFileDialog)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:/a\zim-desktop-wiki\zim-desktop-wiki/tests/widgets.py", line 415, in runTest
    self.assertEqual(myfolder.uri, folder.uri)
AssertionError: 'file[13 chars]desktop-wiki/zim-desktop-wiki/tests/tmp/TestFileDialog' != 'file[13 chars]desktop-wiki/zim-desktop-wiki/tests/tmp/TestFileDialog/folder1'
- file:///D:/a/zim-desktop-wiki/zim-desktop-wiki/tests/tmp/TestFileDialog
+ file:///D:/a/zim-desktop-wiki/zim-desktop-wiki/tests/tmp/TestFileDialog/folder1
?                                                                        ++++++++


----------------------------------------------------------------------
Ran 889 tests in 38.343s
```

See https://github.com/pdecat/zim-desktop-wiki/pull/1/checks?check_run_id=3743069136#step:12:1214 for python 3.8
and https://github.com/pdecat/zim-desktop-wiki/pull/1/checks?check_run_id=3743069179#step:12:1201 for python 3.9
and https://github.com/pdecat/zim-desktop-wiki/pull/1/checks?check_run_id=3743069220#step:12:1201 for python 3.10-dev

# MacOS

```
======================================================================
ERROR: testMoveCaseSensitive (tests.newfs.TestLocalFS)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/newfs.py", line 563, in testMoveCaseSensitive
    re = file.moveto(root.file('FOO.txt'))
  File "/Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/zim/newfs/local.py", line 130, in moveto
    raise FileExistsError(other)
zim.newfs.base.FileExistsError: File or folder already exists: /Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/tmp/TestLocalFS_testMoveCaseSensitive_testMoveCaseSensitive/FOO.txt

======================================================================
FAIL: testImageFile (tests.newfs.TestLocalFS)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/newfs.py", line 791, in testImageFile
    self.assertIn(file.mimetype(), ('image/png', 'image/x-png'))
AssertionError: Lists differ: [':žžž', ':Bar', ':Maßnahmen', ':Placeholder', ':Test'] != [':Bar', ':Maßnahmen', ':Placeholder', ':Test', ':žžž']

First differing element 0:
':žžž'
':Bar'

- [':žžž', ':Bar', ':Maßnahmen', ':Placeholder', ':Test']
?  --------

+ [':Bar', ':Maßnahmen', ':Placeholder', ':Test', ':žžž']
?                                               ++++++++


======================================================================
FAIL: runTest (tests.widgets.TestPageEntry)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/widgets.py", line 229, in runTest
    self.assertEqual([r[0] for r in self.completion_model], wanted)
AssertionError: Lists differ: [':žžž', ':Bar', ':Maßnahmen', ':Placeholder', ':Test'] != [':Bar', ':Maßnahmen', ':Placeholder', ':Test', ':žžž']

First differing element 0:
':žžž'
':Bar'

- [':žžž', ':Bar', ':Maßnahmen', ':Placeholder', ':Test']
?  --------

+ [':Bar', ':Maßnahmen', ':Placeholder', ':Test', ':žžž']
?                                               ++++++++


======================================================================
FAIL: testAllMenuItemsAreDocumented (tests.mainwindow.TestMenuDocs)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/mainwindow.py", line 285, in testAllMenuItemsAreDocumented
    self.assertTrue(label in self.manual, 'Menu item "{}" not documented'.format(label))
AssertionError: False is not true : Menu item "**New Page <⌘N>**" not documented

----------------------------------------------------------------------
Ran 889 tests in 52.852s
```

See https://github.com/pdecat/zim-desktop-wiki/pull/1/checks?check_run_id=3743068778#step:10:1142 for python 3.8
and https://github.com/pdecat/zim-desktop-wiki/pull/1/checks?check_run_id=3743068870#step:10:1119 for python 3.9
and https://github.com/pdecat/zim-desktop-wiki/pull/1/checks?check_run_id=3743068930#step:10:1119 for python 3.10-dev

**Comment by pdecat** [comment]
Rebased on develop to get newfs fixes from https://github.com/zim-desktop-wiki/zim-desktop-wiki/issues/107

**Comment by pdecat** [comment]
# Current MacOS failures
```
======================================================================
ERROR: testMoveCaseSensitive (tests.newfs.TestLocalFS)
----------------------------------------------------------------------
Could not run desktop-file-validate
Traceback (most recent call last):

  File "/Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/newfs.py", line 563, in testMoveCaseSensitive


    re = file.moveto(root.file('FOO.txt'))

  File "/Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/zim/newfs/local.py", line 130, in moveto

    raise FileExistsError(other)
Could not find an application for side-by-side comparison
zim.newfs.base.FileExistsError: File or folder already exists: /Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/tmp/TestLocalFS_testMoveCaseSensitive_testMoveCaseSensitive/FOO.txt

======================================================================
FAIL: testImageFile (tests.newfs.TestLocalFS)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/newfs.py", line 791, in testImageFile
    self.assertIn(file.mimetype(), ('image/png', 'image/x-png'))
AssertionError: 'text/plain' not found in ('image/png', 'image/x-png')

======================================================================
FAIL: testGetMimeType (tests.applications.TestApplicationManager)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/applications.py", line 151, in testGetMimeType
    self.assertEqual(get_mimetype(obj), mimetype)
AssertionError: 'text/plain' != 'image/png'
- text/plain
+ image/png



======================================================================
FAIL: runTest (tests.applications.TestXDGMimeInfo)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/applications.py", line 36, in runTest
    self.assertIsInstance(icon, GdkPixbuf.Pixbuf)
AssertionError: None is not an instance of <class 'gi.overrides.GdkPixbuf.Pixbuf'>

======================================================================
FAIL: testAllMenuItemsAreDocumented (tests.mainwindow.TestMenuDocs)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/mainwindow.py", line 285, in testAllMenuItemsAreDocumented
    self.assertTrue(label in self.manual, 'Menu item "{}" not documented'.format(label))
AssertionError: False is not true : Menu item "**New Page <⌘N>**" not documented

----------------------------------------------------------------------
Ran 892 tests in 74.669s

FAILED (failures=4, errors=1, skipped=29, expected failures=21)
```

See https://github.com/pdecat/zim-desktop-wiki/pull/1/checks?check_run_id=3863838739#step:10:1040
and https://github.com/pdecat/zim-desktop-wiki/pull/1/checks?check_run_id=3863838789#step:10:1047
and https://github.com/pdecat/zim-desktop-wiki/pull/1/checks?check_run_id=3863838831#step:10:1039

**Comment by pdecat** [comment]
Rebased on current develop branch.

**Comment by jaap-karssenberg** [comment]
Thanks for keeping it warm - have it high on my list to start digging in
these test suite issues - just no spare time last weeks.

**Comment by dehesselle** [comment]
> `ERROR: testMoveCaseSensitive (tests.newfs.TestLocalFS)`

macOS filesystems are case insensitive by default. The failure of that test is expected behavior.

**Comment by jaap-karssenberg** [comment]
@dehesselle what should we use for mimetypes on mac OS - can we assume xdg.Mime to be available ?

Background is that several tests fail due to mimetype errors. Seems xdg.Mime is installed, but no database present.

**Comment by jaap-karssenberg** [comment]
@pdecat merging 594528dc1dabfd6be5b80a885736786aa49383ee should take care of test failures on windows

**Comment by pdecat** [comment]
Rebased on current develop branch.

**Comment by pdecat** [comment]
Tests are now all passing on Windows too 🎉

**Comment by jaap-karssenberg** [comment]
Cool !  I'll look into the remaining mac OS issues tonight.

**Comment by dehesselle** [comment]
Do you want me to look into the mime thing @jaap-karssenberg ? I haven't said anything yesterday because I had no update on this and wanted to avoid unnecessary noise in the issue. I can probably look into it end of the week.

**Comment by jaap-karssenberg** [comment]
Thanks but I realized I can just detect the database without needing to assume anything. Will submit patch later.

**Comment by jaap-karssenberg** [comment]
With a bit of luck 327736b8e1b2c9777116ca425fe496aa8b7cc935 should let mac OS pass as well

**Comment by pdecat** [comment]
I've rebased to include https://github.com/zim-desktop-wiki/zim-desktop-wiki/commit/327736b8e1b2c9777116ca425fe496aa8b7cc935 but there are still a few failures on MacOS:

```
======================================================================
ERROR: testMoveCaseSensitive (tests.newfs.TestLocalFS)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/newfs.py", line 563, in testMoveCaseSensitive
    re = file.moveto(root.file('FOO.txt'))
  File "/Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/zim/newfs/local.py", line 130, in moveto
    raise FileExistsError(other)
zim.newfs.base.FileExistsError: File or folder already exists: /Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/tmp/TestLocalFS_testMoveCaseSensitive_testMoveCaseSensitive/FOO.txt

======================================================================
FAIL: runTest (tests.applications.TestXDGMimeInfo)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/applications.py", line 36, in runTest
    self.assertIsInstance(icon, GdkPixbuf.Pixbuf)
AssertionError: None is not an instance of <class 'gi.overrides.GdkPixbuf.Pixbuf'>

======================================================================
FAIL: testAllMenuItemsAreDocumented (tests.mainwindow.TestMenuDocs)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/mainwindow.py", line 292, in testAllMenuItemsAreDocumented
    self.assertTrue(label in self.manual, 'Menu item "{}" not documented'.format(label))
AssertionError: False is not true : Menu item "**New Sub Page <⇧Ctrl><N>**" not documented

----------------------------------------------------------------------
Ran 910 tests in 64.084s

FAILED (failures=2, errors=1, skipped=29, expected failures=21)
```

**Comment by jaap-karssenberg** [comment]
Try 2: 9c18b13b4779b86af1e28f45a22d176ba970744d

**Comment by jaap-karssenberg** [comment]
Btw. sorry for testing it via your branch, but would look to be all tests green when we pull in the main archive - makes the overview more clean going forward

**Comment by pdecat** [comment]
Only one failure left on MacOS:

```
======================================================================
FAIL: testAllMenuItemsAreDocumented (tests.mainwindow.TestMenuDocs)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/zim-desktop-wiki/zim-desktop-wiki/tests/mainwindow.py", line 293, in testAllMenuItemsAreDocumented
    self.assertTrue(label in self.manual, 'Menu item "{}" not documented'.format(label))
AssertionError: False is not true : Menu item "**Back <⌥←>**" not documented

----------------------------------------------------------------------
Ran 910 tests in 59.316s

FAILED (failures=1, skipped=30, expected failures=21)
```

**Comment by jaap-karssenberg** [comment]
Merged just now - fingers crossed ;)

**Comment by pdecat** [comment]
🥁
https://github.com/zim-desktop-wiki/zim-desktop-wiki/actions/runs/1496396784

**Comment by pdecat** [comment]
There's one new failure on windows but only with python 3.9 🤔

**Comment by jaap-karssenberg** [comment]
I'll disable that one - seen it fail in the past from time to time at
randomly - there is some race condition hidden in the Gtk file dialog

**Comment by jaap-karssenberg** [comment]
Mac OS is now also passing - although one run seems to be hanging

**Pull request closed without merging**